// apollo-datasource-rest/dist/RESTDataSource.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
const apollo_server_errors_1 = require("apollo-server-errors");
class RESTDataSource {
    didReceiveErrorResponse(response) {
        return __awaiter(this, void 0, void 0, function* () {
            const message = `${response.status} ${response.statusText}: ${yield response.text()}`;
            if (response.status === 401) {
                throw new apollo_server_errors_1.AuthenticationError(message);
            }
            else if (response.status === 403) {
                throw new apollo_server_errors_1.ForbiddenError(message);
            }
            else {
                throw new apollo_server_errors_1.ApolloError(message);
            }
        });
    }
    get(path, params, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.fetch(path, Object.assign({ method: 'GET', params }, options));
        });
    }
    post(path, body, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.fetch(path, Object.assign({ method: 'POST', body }, options));
        });
    }
    patch(path, body, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.fetch(path, Object.assign({ method: 'PATCH', body }, options));
        });
    }
    put(path, body, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.fetch(path, Object.assign({ method: 'PUT', body }, options));
        });
    }
    delete(path, params, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.fetch(path, Object.assign({ method: 'DELETE', params }, options));
        });
    }
    fetch(path, options) {
        return __awaiter(this, void 0, void 0, function* () {
            const { params } = options, init = __rest(options, ["params"]);
            const normalizedBaseURL = this.baseURL.endsWith('/')
                ? this.baseURL
                : this.baseURL.concat('/');
            const url = new URL(path, normalizedBaseURL);
            if (params) {
                for (const [name, value] of new URLSearchParams(params)) {
                    url.searchParams.append(name, value);
                }
            }
            if (init.body !== undefined &&
                typeof init.body !== 'string' &&
                !(init.body instanceof ArrayBuffer)) {
                init.body = JSON.stringify(init.body);
                if (!(init.headers instanceof Headers)) {
                    init.headers = new Headers(init.headers);
                }
                init.headers.set('Content-Type', 'application/json');
            }
            return this.trace(`${init.method || 'GET'} ${url}`, () => __awaiter(this, void 0, void 0, function* () {
                const request = new Request(String(url), init);
                if (this.willSendRequest) {
                    this.willSendRequest(request);
                }
                const response = yield this.httpCache.fetch(request);
                if (response.ok) {
                    const contentType = response.headers.get('Content-Type');
                    if (contentType && contentType.startsWith('application/json')) {
                        return response.json();
                    }
                    else {
                        return response.text();
                    }
                }
                else {
                    return this.didReceiveErrorResponse(response);
                }
            }));
        });
    }
    trace(label, fn) {
        return __awaiter(this, void 0, void 0, function* () {
            if (process && process.env && process.env.NODE_ENV === 'development') {
                const startTime = Date.now();
                try {
                    return yield fn();
                }
                finally {
                    const duration = Date.now() - startTime;
                    console.log(`${label} (${duration}ms)`);
                }
            }
            else {
                return fn();
            }
        });
    }
}
exports.RESTDataSource = RESTDataSource;
